packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx

```diff
@@ -0,0 +1,302 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { GrafanaTheme2, PanelMenuItem } from '@grafana/data';
+import { css, cx } from '@emotion/css';
+import { useStyles2 } from '../../themes';
+import { PanelMenuListItem } from './PanelMenuItem';
+import { Icon } from '../Icon/Icon';
+import { ClickOutsideWrapper } from '../ClickOutsideWrapper/ClickOutsideWrapper';
+
+interface Props {
+  items: PanelMenuItem[];
+  title: string;
+  buttonClassName?: string;
+  outside?: React.ReactElement;
+  inside?: React.ReactElement;
+}
+
+const generateMarkup = (item: PanelMenuItem, styles: ReturnType<typeof panelMenuStyles>, items: PanelMenuItem[]) => {
+  if (item.type === 'divider') {
+    return <li id="divider" role="none" className={styles.divider} key="divider"></li>;
+  }
+
+  if (item.subMenu && item.subMenu?.length > 0) {
+    return (
+      <PanelMenuListItem key={item.text} item={item}>
+        <ul className={cx(styles.panelMenu, styles.subMenu)} role="menu" id="panel-menu">
+          {item.subMenu.map((i) => generateMarkup(i, styles, items))}
+        </ul>
+      </PanelMenuListItem>
+    );
+  } else {
+    return <PanelMenuListItem key={item.text} item={item} />;
+  }
+};
+
+export const PanelMenu = ({ items, title, buttonClassName, outside, inside }: Props) => {
+  const [isMouseDown, setIsMouseDown] = useState(false);
+  const [isDragging, setIsDragging] = useState(false);
+  const onMouseDown = useCallback(() => {
+    setIsMouseDown(true);
+    setIsDragging(false);
+  }, []);
+
+  const onMouseUp = useCallback(() => {
+    setIsMouseDown(false);
+  }, []);
+
+  const onMouseMove = useCallback(() => {
+    if (isMouseDown && !isDragging) {
+      setIsDragging(true);
+    }
+  }, [isMouseDown, isDragging]);
+
+  const [menuIsOpen, setMenuOpen] = useState(false);
+  const toggleMenu = useCallback(
+    (event: React.SyntheticEvent) => {
+      event.preventDefault();
+      if (isDragging && !menuIsOpen) {
+        return;
+      }
+
+      setMenuOpen(!menuIsOpen);
+    },
+    [menuIsOpen, setMenuOpen, isDragging]
+  );
+
+  const handleKeys = (event: React.KeyboardEvent) => {
```

Comment:
I want to refactor all this key-handling logic, and move it to a separate file maybe. Pretty messy and repetitive at the moment.